cartographer/common/rate_timer.h

```diff
@@ -0,0 +1,134 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_COMMON_RATE_TIMER_H_
+#define CARTOGRAPHER_COMMON_RATE_TIMER_H_
+
+#include <chrono>
+#include <deque>
+#include <iomanip>
+#include <numeric>
+#include <sstream>
+#include <string>
+#include <vector>
+
+#include "cartographer/common/port.h"
+#include "cartographer/common/time.h"
+
+namespace cartographer {
+namespace common {
+
+// Computes the rate at which pulses come in.
+template <typename ClockType = std::chrono::steady_clock>
+class RateTimer {
+ public:
+  // Computes the rate at which pulses come in over 'window_duration' in wall
+  // time.
+  explicit RateTimer(const common::Duration window_duration)
+      : window_duration_(window_duration) {}
+  ~RateTimer() {}
+
+  RateTimer(const RateTimer&) = delete;
+  RateTimer& operator=(const RateTimer&) = delete;
+
+  // Returns the pulse rate in Hz.
+  double ComputeRate() const {
+    if (events_.empty()) {
+      return 0.;
+    }
+    return static_cast<double>(events_.size() - 1) /
+           common::ToSeconds((events_.back().time - events_.front().time));
+  }
+
+  // Returns the ratio of the pulse rate (with supplied times) to the wall time
+  // rate. For example, if a sensor produces pulses at 10 Hz, but we call Pulse
+  // at 20 Hz wall time, this will return 2.
+  double ComputeWallTimeRateRatio() const {
+    if (events_.empty()) {
+      return 0.;
+    }
+    return common::ToSeconds((events_.back().time - events_.front().time)) /
+           std::chrono::duration_cast<std::chrono::duration<double>>(
+               events_.back().wall_time - events_.front().wall_time)
+               .count();
+  }
+
+  // Records an event that will contribute to the computed rate.
+  void Pulse(common::Time time) {
+    events_.push_back(Event{time, ClockType::now()});
+    while (events_.size() > 2 &&
+           (events_.back().wall_time - events_.front().wall_time) >
+               window_duration_) {
+      events_.pop_front();
+    }
+  }
+
+  // Computes all differences in seconds between consecutive pulses.
+  std::vector<double> ComputeDeltasInSeconds() const {
+    if (events_.size() < 2) {
+      return {};
+    }
+    const size_t count = events_.size() - 1;
+    std::vector<double> result;
+    result.reserve(count);
+    for (size_t i = 0; i != count; ++i) {
+      result.push_back(
+          common::ToSeconds(events_[i + 1].time - events_[i].time));
+    }
+    return result;
+  }
+
+  // Returns the average and standard deviation of the deltas.
+  string DeltasDebugString() const {
+    const auto deltas = ComputeDeltasInSeconds();
+    const double sum = std::accumulate(deltas.begin(), deltas.end(), 0.);
+    const double mean = sum / deltas.size();
+
+    std::vector<double> diff(deltas.size());
```

Comment:
This could be rewritten similar to:
double squared_sum = 0.;
for (const double x : deltas) {
  squared_sum += common::Pow2(x - mean);
}

This would not only be more concise and readable, but also save memory.